### ISSUE TITLE
moq-relay-ietf crashes when running in Mininet environment

### ISSUE THREAD

**Issue by RuonanChai** [moq-rs/moq#2]
Hi everyone,

I am trying to run moq-relay-ietf in a virtual network environment created by Mininet, but I encountered two serious bugs.

Environment:

 - OS: Ubuntu 22.04 

 - Virtualization: Mininet

Topology: A multi-tier tree topology. (I will attach the Python script that generates the network to this issue)

Steps to Reproduce:

1. Use the attached Python script to start the Mininet network.

2. In the r0 node of Mininet, try to start moq-relay-ietf.

Observed Bugs:

Bug 1: When binding to a specific IP, the program crashes after receiving the first connection.

Command: moq-relay-ietf --bind 10.0.1.1:5443 --tls-cert cert.pem --tls-key key.pem

Phenomenon: The program starts fine. But when a client (such as curl -k https://10.0.1.1:5443) tries to connect, the program immediately crashes.

Bug 2: When binding to 0.0.0.0, the program fails silently and does not listen to any ports.

Command: moq-relay-ietf --bind 0.0.0.0:5443 --tls-cert cert.pem --tls-key key.pem

Phenomenon: The program does not crash after starting, but it does not actually listen to any ports. Checking with the netstat -lntp command shows that no process is listening on port 5443. Therefore, the client receives a "Connection refused" error when connecting.

Expected Result:
No matter in which binding mode, moq-relay-ietf should start successfully, listen on the specified port, and accept client connections normally without crashing.

Attachment:
import os
import sys
import time
import argparse
import traceback
import subprocess
from mininet.net import Mininet
from mininet.topo import Topo
from mininet.link import TCLink
from mininet.log import setLogLevel, info, error
from mininet.node import OVSKernelSwitch
from mininet.cli import CLI

CURRENT_DIR = os.path.dirname(os.path.abspath(__file__))
REGIONAL_CONTROLLER_SCRIPT_PATH = os.path.join(CURRENT_DIR, "regional_relay_controller.py")
TCP_BASELINE_SCRIPT_PATH = os.path.join(CURRENT_DIR, "run_tcp_baseline.py")

class MultiTierTopo(Topo):
    def build(self, num_users=20):
        s1 = self.addSwitch('s1')
        s2 = self.addSwitch('s2')
        s_edges = [self.addSwitch(f's1{i}') for i in range(4)]

        source = self.addHost('n0', ip='10.0.1.100/24')
        r0 = self.addHost('r0')
        r1 = self.addHost('r1')
        r2 = self.addHost('r2')
        edge_relays = [self.addHost(f'r{i}') for i in range(3, 7)]

        self.addLink(source, s1, bw=1000)
        self.addLink(r0, s1, bw=1000)
        self.addLink(r0, s2, bw=500)
        self.addLink(r1, s2, bw=500)
        self.addLink(r2, s2, bw=500)
        self.addLink(r1, s_edges[0], bw=100)
        self.addLink(r1, s_edges[1], bw=100)
        self.addLink(r2, s_edges[2], bw=100)
        self.addLink(r2, s_edges[3], bw=100)
        for i, er in enumerate(edge_relays):
            self.addLink(er, s_edges[i], bw=100)
        for i in range(1, num_users + 1):
            edge_relay_index = (i - 1) % 4
            switch = s_edges[edge_relay_index]
            subnet = 3 + edge_relay_index
            host_ip = f"10.0.{subnet}.{10 + i}/24"
            host = self.addHost(f"h{i}", ip=host_ip)
            bw = [4, 8, 15][(i - 1) % 3]
            self.addLink(host, switch, bw=bw, delay="5ms", loss=0.1)

def setup_host_interfaces(net, users):
    # 必须按接口顺序设置IP
    r0 = net.get('r0')
    r1 = net.get('r1')
    r2 = net.get('r2')
    n0 = net.get('n0')
   # 打开转发
    for relay_name in ['r0', 'r1', 'r2', 'r3', 'r4', 'r5', 'r6']:
        net.get(relay_name).cmd('sysctl -w net.ipv4.ip_forward=1')

    # 先清除所有接口IP，防止Mininet自动分配有冲突
    for node in [r0, r1, r2, n0]:
        for i in range(len(node.intfList())):
            node.cmd(f'ifconfig {node.name}-eth{i} 0')

    r0.setIP('10.0.1.1/24', intf='r0-eth0')
    r0.setIP('10.0.2.254/24', intf='r0-eth1')
    r1.setIP('10.0.2.1/24', intf='r1-eth0')
    r1.setIP('10.0.3.1/24', intf='r1-eth1')
    r1.setIP('10.0.4.1/24', intf='r1-eth2')
    r2.setIP('10.0.2.2/24', intf='r2-eth0')
    r2.setIP('10.0.5.1/24', intf='r2-eth1')
    r2.setIP('10.0.6.1/24', intf='r2-eth2')
    n0.setIP('10.0.1.100/24', intf='n0-eth0')

    net.get('r3').setIP('10.0.3.2/24', intf='r3-eth0')   # 注意不能和r1冲突
    net.get('r4').setIP('10.0.4.2/24', intf='r4-eth0')
    net.get('r5').setIP('10.0.5.2/24', intf='r5-eth0')
    net.get('r6').setIP('10.0.6.2/24', intf='r6-eth0')

    # 正确设置默认网关（必须指向regional relay接口，不是自己！）
    net.get('r3').cmd('route add default gw 10.0.3.1')  # r1的10.0.3.1
    net.get('r4').cmd('route add default gw 10.0.4.1')  # r1的10.0.4.1
    net.get('r5').cmd('route add default gw 10.0.5.1')  # r2的10.0.5.1
    net.get('r6').cmd('route add default gw 10.0.6.1')  # r2的10.0.6.1

    # 设置r1/r2的默认网关
    net.get('r1').cmd('route add default gw 10.0.2.254')
    net.get('r2').cmd('route add default gw 10.0.2.254')

    # n0 配路由到所有区域
    for subnet in ['10.0.2.0/24','10.0.3.0/24','10.0.4.0/24','10.0.5.0/24','10.0.6.0/24']:
        n0.cmd(f'route add -net {subnet} gw 10.0.1.1')

    # r0 到各区域的路由
    r0.cmd('route add -net 10.0.3.0/24 gw 10.0.2.1')
    r0.cmd('route add -net 10.0.4.0/24 gw 10.0.2.1')
    r0.cmd('route add -net 10.0.5.0/24 gw 10.0.2.2')
    r0.cmd('route add -net 10.0.6.0/24 gw 10.0.2.2')

    # 用户主机默认网关仍指向其edge relay
    for i in range(1, users + 1):
        h = net.get(f"h{i}")
        edge_subnet_prefix = 3 + ((i-1)%4)
        gateway_ip = f"10.0.{edge_subnet_prefix}.2"  # 指向r3~r6的IP
        h.cmd(f'route add default gw {gateway_ip}')

    for sw in net.switches:
        sw.cmd(f'ovs-ofctl add-flow {sw.name} priority=1,actions=NORMAL')

def auto_check_links(net, users, log_path=None):
    '''
    检查所有主机到edge relay、regional relay、root relay、n0的网络连通性
    '''
    targets = ['r3', 'r4', 'r5', 'r6', 'r1', 'r2', 'r0', 'n0']
    summary = []
    info("\n========== [AUTO LINK CHECK] ==========\n")
    for i in range(1, users+1):
        h = net.get(f'h{i}')
        for relay in targets:
            dst_ip = net.get(relay).IP()
            out = h.cmd(f'ping -c 1 -w 1 {dst_ip}')
            ok = '1 received' in out or '1 packets transmitted, 1 received' in out
            msg = f"h{i} ({h.IP()}) --> {relay} ({dst_ip}): {'OK' if ok else 'FAIL'}"
            info(msg + '\n')
            summary.append(msg)
    # 写入链路检查log
    if log_path:
        with open(os.path.join(log_path, "auto_link_check.log"), "w") as f:
            for line in summary:
                f.write(line + "\n")
    info("========== [AUTO LINK CHECK END] ==========\n\n")

def start_clients_in_batches(net, num_users, command_gen, log_path=None, batch_size=5):
    info(f'*** Starting {num_users} clients in batches of {batch_size}...\n')
    all_processes = []
    for i in range(1, num_users + 1, batch_size):
        for j in range(i, min(i+batch_size, num_users+1)):
            h = net.get(f"h{j}")
            cmd = command_gen(j)
            # stdout, stderr统一重定向
            outlog = os.path.join(log_path, f"h{j}_output.log") if log_path else "/dev/null"
            p = h.popen(cmd, shell=True, stdout=open(outlog,"w"), stderr=subprocess.STDOUT)
            all_processes.append(p)
        time.sleep(1)
    info(f'*** All {num_users} client processes have been launched. Waiting for completion... ***\n')
    for p in all_processes:
        p.wait()
        if p.returncode != 0:
            info(f"[WARNING] 客户端进程以错误码 {p.returncode} 退出。\n")

def run_moq_scenario(net, args):
    info("*** 配置 MoQ (MD2G) 场景...\n")
    source_host = net.get('n0')
    root_relay = net.get('r0')
    regional_relays = [net.get('r1'), net.get('r2')]
    all_relays = [root_relay] + regional_relays + [net.get(f'r{i}') for i in range(3, 7)]
    # 启动 moq-relay 服务
    for r in all_relays:
        r.popen(
        [
            args.relay_bin,
            "--bind", f"{r.IP()}:5443",
            "--tls-cert", f"{args.cert_dir}/cert.pem",
            "--tls-key", f"{args.cert_dir}/key.pem"
        ],
        stdout=open(f"/tmp/moq-relay-{r.name}.log", "w"),
        stderr=subprocess.STDOUT
    )
    time.sleep(3)
    for relay_host in all_relays:
        port = 5443
        cmd = f'netstat -lnt | grep :{port}'
        result = relay_host.cmd(cmd)
        retry_count = 0
        while f":{port} " not in result and retry_count < 5:
            info(f"  - Relay on {relay_host.name} not up yet. Retrying...\n")
            time.sleep(1)
            result = relay_host.cmd(cmd)
            retry_count += 1
        if f":{port} " not in result:
            info(f"  - [FATAL] Relay on {relay_host.name} 未能在端口 {port} 启动监听，实验中止。\n")
            return 
    info('*** 所有 moq-relay 进程已启动。\n')
    for stream_type in ["base", "enhanced"]:
        video_file = os.path.join(args.video_dir, f"redandblack_{stream_type}_avc_streamable.mp4")
        if os.path.exists(video_file):
            pub_cmd = (f'nohup ffmpeg -re -stream_loop -1 -i {video_file} -c copy -f data - | '
                       f'{args.pub_bin} --name redandblack/{stream_type} https://{root_relay.IP()}:5443 > '
                       f'/tmp/moq-pub-{stream_type}.log 2>&1 &')
            source_host.cmd(pub_cmd)
    time.sleep(2)
    info("*** moq-pub 推流已启动。\n")
    users_per_regional = args.users // 2
    decision_file_r1 = f"/tmp/r1_decisions.json"
    r1_cmd = (f'nohup {sys.executable} {REGIONAL_CONTROLLER_SCRIPT_PATH} '
              f'--model_path {args.model_path} --num_users {args.users} --user_offset 0 '
              f'--decision_file {decision_file_r1} > /tmp/controller_r1.log 2>&1 &')
    regional_relays[0].cmd(r1_cmd)
    decision_file_r2 = f"/tmp/r2_decisions.json"
    r2_cmd = (f'nohup {sys.executable} {REGIONAL_CONTROLLER_SCRIPT_PATH} '
              f'--model_path {args.model_path} --num_users {args.users} --user_offset {users_per_regional} '
              f'--decision_file {decision_file_r2} > /tmp/controller_r2.log 2>&1 &')
    regional_relays[1].cmd(r2_cmd)
    info("*** PPO 区域控制器已启动。\n")
    time.sleep(2)
    def moq_command_generator(host_id):
        edge_relay_index = (host_id - 1) % 4
        regional_relay_name = 'r1' if edge_relay_index < 2 else 'r2'
        decision_file_path = decision_file_r1 if regional_relay_name == 'r1' else decision_file_r2
        edge_relay_ip = net.get(f"r{3+edge_relay_index}").IP()
        log_file = os.path.join(args.log_path, f"client_h{host_id}_SCRIPT_ERROR.log")
        return (f'{sys.executable} {args.dispatch_script} '
                f'--host_id {host_id} --log_path {args.log_path} --strategy {args.strategy} '
                f'--decision_file {decision_file_path} --users {args.users} '
                f'--relay_ip {edge_relay_ip} ')
    start_clients_in_batches(net, args.users, moq_command_generator, args.log_path)

def run_tcp_baseline_scenario(net, args):
    info(f"*** 配置 TCP Baseline ({args.strategy}) 场景...\n")
    edge_relays = [net.get(f'r{i}') for i in range(3, 7)]
    for r in edge_relays:
        server_cmd = (f'nohup {sys.executable} {TCP_BASELINE_SCRIPT_PATH} '
                      f'--role server --bind_ip {r.IP()} > /tmp/tcp_server_{r.name}.log 2>&1 &')
        r.cmd(server_cmd)
    time.sleep(2)
    info("*** 所有 TCP Baseline 服务器已启动。\n")
    def tcp_command_generator(host_id):
        edge_relay_index = (host_id - 1) % 4
        edge_relay = edge_relays[edge_relay_index]
        log_file = os.path.join(args.log_path, f"tcp_client_h{host_id}_SCRIPT_ERROR.log")
        return (f'{sys.executable} {TCP_BASELINE_SCRIPT_PATH} '
                f'--role client --host_id {host_id} --log_path {args.log_path} '
                f'--server_ip {edge_relay.IP()} ')
    start_clients_in_batches(net, args.users, tcp_command_generator, args.log_path)

def run_experiment(args):
    os.system("sudo mn -c >/dev/null 2>&1")
    info(f"*** Creating log directory: {args.log_path}\n")
    os.makedirs(args.log_path, exist_ok=True)
    topo = MultiTierTopo(num_users=args.users)
    net = Mininet(topo=topo, link=TCLink, controller=None, switch=OVSKernelSwitch, autoSetMacs=True)
    info("*** Starting Mininet network...\n")
    net.start()
    info("*** Configuring host IP/routing...\n")
    setup_host_interfaces(net, args.users)
    auto_check_links(net, args.users, args.log_path)
    info("*** Waiting for network to stabilize...\n")
    time.sleep(2)
    # info("*** Pinging all hosts to check connectivity...\n")
    # loss = net.pingAll(timeout=1)
    # if loss > 0:
    #     error(f"*** Network NOT fully connected! Packet loss: {loss}%\n")
    # else:
    #     info("*** SUCCESS: Network fully connected.\n")
    CLI(net)
    try:
        if not args.baseline_tcp:
            run_moq_scenario(net, args)
        else:
            run_tcp_baseline_scenario(net, args)
    except Exception as e:
        error(f"*** An exception occurred during the experiment: {e}\n")
        traceback.print_exc()
    finally:
        os.system("pkill -9 -f moq-relay; pkill -9 -f moq-pub; pkill -9 -f ffmpeg; "
                  "pkill -9 -f regional_relay_controller.py; pkill -9 -f run_tcp_baseline.py; "
                  "pkill -9 -f dispatch_strategy.py >/dev/null 2>&1")
        net.stop()
        info("*** Network stopped.\n")

if __name__ == '__main__':
    setLogLevel('info')
    parser = argparse.ArgumentParser(description="Mininet Experiment Runner")
    parser.add_argument('--strategy', type=str, required=True)
    parser.add_argument('--users', type=int, required=True)
    parser.add_argument('--log_path', type=str, required=True)
    parser.add_argument('--model_path', type=str)
    parser.add_argument('--relay_bin', type=str)
    parser.add_argument('--pub_bin', type=str)
    parser.add_argument('--cert_dir', type=str)
    parser.add_argument('--video_dir', type=str)
    parser.add_argument('--dispatch_script', type=str)
    parser.add_argument('--baseline_tcp', action='store_true')
    args = parser.parse_args()
    if not args.baseline_tcp:
        required_for_moq = ['model_path', 'relay_bin', 'pub_bin', 'cert_dir', 'video_dir', 'dispatch_script']
        missing = [arg for arg in required_for_moq if getattr(args, arg) is None]
        if missing:
            parser.error(f"对于MoQ策略, 以下参数是必需的: --{', --'.join(missing)}")
    run_experiment(args)
